src/detect-app-layer-state.c

```diff
@@ -0,0 +1,229 @@
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Victor Julien <vjulien@oisf.net>
+ */
+
+#include "suricata-common.h"
+#include "threads.h"
+#include "decode.h"
+
+#include "app-layer.h"
+#include "app-layer-protos.h"
+#include "app-layer-parser.h"
+#include "app-layer-smtp.h"
+#include "detect.h"
+#include "detect-parse.h"
+#include "detect-engine.h"
+#include "detect-engine-state.h"
+#include "detect-engine-build.h"
+#include "detect-app-layer-state.h"
+
+#include "flow.h"
+#include "flow-var.h"
+#include "flow-util.h"
+
+#include "decode-events.h"
+#include "util-byte.h"
+#include "util-debug.h"
+#include "util-enum.h"
+#include "util-profiling.h"
+#include "util-unittest.h"
+#include "util-unittest-helper.h"
+#include "stream-tcp-util.h"
+
+typedef struct DetectAppLayerStateData_ {
+    uint8_t progress;
+    int8_t mode;
+} DetectAppLayerStateData;
+
+static int DetectAppLayerStateSetup(DetectEngineCtx *, Signature *, const char *);
+static void DetectAppLayerStateFree(DetectEngineCtx *, void *);
+static uint8_t DetectEngineAptStateInspect(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const struct DetectEngineAppInspectionEngine_ *engine, const Signature *s, Flow *f,
+        uint8_t flags, void *alstate, void *tx, uint64_t tx_id);
+static int g_applayer_state_list_id = 0;
+
+void DetectAppLayerStateRegister(void)
+{
+    sigmatch_table[DETECT_APP_LAYER_STATE].name = "app-layer-state";
+    sigmatch_table[DETECT_APP_LAYER_STATE].desc =
+            "match on events generated by the App Layer Parsers and the protocol detection engine";
+    sigmatch_table[DETECT_APP_LAYER_STATE].url = "/rules/app-layer.html#app-layer-event";
+    sigmatch_table[DETECT_APP_LAYER_STATE].Setup = DetectAppLayerStateSetup;
+    sigmatch_table[DETECT_APP_LAYER_STATE].Free = DetectAppLayerStateFree;
+
+    DetectAppLayerInspectEngineRegister("app-layer-state", ALPROTO_UNKNOWN, SIG_FLAG_TOSERVER, 0,
+            DetectEngineAptStateInspect, NULL);
+    DetectAppLayerInspectEngineRegister("app-layer-state", ALPROTO_UNKNOWN, SIG_FLAG_TOCLIENT, 0,
+            DetectEngineAptStateInspect, NULL);
+
+    g_applayer_state_list_id = DetectBufferTypeGetByName("app-layer-state");
+}
+
+static uint8_t DetectEngineAptStateInspect(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
```

Comment:
can try make this a regular keyword now that the request_update logic is removed